### PR TITLE
[TRTLLM-12669][refactor] Eagle3 sampling: auto-detect greedy fast-path, mixed-batch rejection sampling, draft honors target params

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
Replace static config flag with auto-detected per-step uses_advanced_sampling based on actual sampling params. Include this in CUDA graph key so we lazily capture two graph variants (argmax fast-path vs advanced sampling kernel) and dispatch by replaying the right one.

## Summary by CodeRabbit

* **New Features**
  * Speculative decoding now automatically detects all-greedy sampling for optimized performance.
  * Draft token generation enhanced to support per-request sampling parameters in advanced configurations.

* **Bug Fixes**
  * Removed unnecessary fallback warnings in speculative decoding setup.

* **Documentation**
  * Updated speculative decoding configuration examples and test settings with improved sampling support.

## Description

Refactors Eagle3 one-model speculative decoding sampling. Four logical changes, sequenced as separate commits:

### 1. Replace `allow_advanced_sampling` with auto-detected `is_all_greedy_sample`

Removes the `allow_advanced_sampling` config flag from `DecodingBaseConfig`. Replaced with a per-step `is_all_greedy_sample` derived from the actual `temperature` / `top_k` / `top_p` of requests in the batch. The flag is included in the CUDA graph cache key, so two graph variants are lazily captured (argmax fast-path vs. advanced-sampling kernel) and dispatched at replay time based on batch composition.

### 2. Eagle3 drafter honors target's sampling params

Previously the Eagle3 draft model always ran greedy regardless of the target's sampling configuration. This change propagates the target's `temperature` / `top_k` / `top_p` into the draft loop so that draft samples come from the **same distribution** as the target's reference distribution. This is a correctness prerequisite for non-greedy rejection sampling (Leviathan formula `u * p_draft(x) < p_target(x)` only holds when both probabilities come from the same conditioning).

### 3. Slot-indexed `draft_probs` to support mixed batches

Previously `_can_use_rejection_sampling` bailed out when the batch contained context requests, falling back to exact-match for the whole batch. Root cause: `draft_probs` was indexed by **batch position**, but batch position is unstable across iterations (chunked-prefill, finishing gens, new ctx joins all shift positions). Fix:

- Reshape `draft_probs` from flat `[total_draft_tokens, vocab]` to slot-indexed `[max_num_requests, max_draft_len, vocab]`, keyed by stable `py_seq_slot`.
- Scatter on write (`_compute_and_store_draft_probs`), gather on read (`_accept_draft_tokens`) using a precomputed `batch_slot_ids` tensor.
- Drop the `num_contexts == 0` constraint in `_can_use_rejection_sampling` — ctx subset goes through `_sample_tokens_for_batch`, gen subset goes through the rejection kernel.
- Reset `draft_probs_valid = False` whenever the draft loop writes no probs, so stale data is never read.

Mixed-batch rejection captures ~18% sys-tps on llama70b bs=32 vs. the exact-match fallback.

---

### Unit tests (B200)

```
tests/unittest/_torch/speculative/test_eagle3.py
  test_eagle3_cuda_graph_padding[True]  PASSED  130.54s
  test_eagle3_cuda_graph_padding[False] PASSED  103.53s
```

### End-to-end correctness — Qwen3-8B (H100 SXM5 80G)

Greedy path verification (no temp / top_p / top_k → both paths take greedy branch):
- baseline (`83ec591`, `allow_advanced_sampling=False`) vs new (`d237690`, auto-detected `is_all_greedy_sample=True`)
- token-level identity: **80/80 prompts identical**
- `total_output_tokens`: 113,432 vs 113,432, Δ=0
- `mean_acceptance_rate`: 0.4918 vs 0.4918, Δ=+0.00%
- `mean_acceptance_length`: 2.475 vs 2.475, Δ=0
- mean throughput: 224.19 vs 224.02 tok/s, Δ=-0.07%
- wall clock: 505.99s vs 506.38s, Δ=+0.08%

## Performance — rejection sampling ON vs OFF (non-greedy)

CUDA graph **enabled**. mtbench dataset. Sampling params: `temperature=0.7, top_k=50, top_p=0.9`.

### Llama-3.3-70B-Instruct + EAGLE-3 (mean over 3 rounds)

| bs | TPS off | TPS on | ΔTPS | AR off | AR on | ΔAR |
|---:|--------:|-------:|-----:|-------:|------:|----:|
| 32 | 3408.2  | 3111.3 | **−8.70%** | 0.3475 | 0.3581 | +3.04% |
| 16 | 2203.7  | 2069.3 | **−6.09%** | 0.3499 | 0.3568 | +1.96% |
|  8 | 1278.9  | 1232.4 | **−3.58%** | 0.3444 | 0.3586 | +4.15% |
|  4 |  704.1  |  694.9 | **−1.29%** | 0.3451 | 0.3585 | +3.88% |
|  2 |  390.2  |  384.3 | **−1.51%** | 0.3448 | 0.3547 | +2.89% |
|  1 |  204.2  |  203.4 | **−0.43%** | 0.3490 | 0.3561 | +2.05% |

<details><summary>Per-round detail (3 rounds, llama-70b)</summary>

| bs | round | TPS off | TPS on | ΔTPS    | AR off | AR on  | ΔAR     |
|---:|:-----:|--------:|-------:|--------:|-------:|-------:|--------:|
| 32 | R1    | 3413.8  | 3113.9 | −8.78%  | 0.3479 | 0.3579 | +2.88%  |
| 32 | R2    | 3441.7  | 3111.2 | −9.60%  | 0.3512 | 0.3584 | +2.04%  |
| 32 | R3    | 3369.0  | 3108.9 | −7.72%  | 0.3435 | 0.3579 | +4.19%  |
| 16 | R1    | 2175.9  | 2078.4 | −4.48%  | 0.3524 | 0.3559 | +0.99%  |
| 16 | R2    | 2219.6  | 2101.3 | −5.33%  | 0.3492 | 0.3591 | +2.82%  |
| 16 | R3    | 2215.7  | 2028.2 | −8.46%  | 0.3481 | 0.3553 | +2.08%  |
|  8 | R1    | 1281.1  | 1230.6 | −3.94%  | 0.3428 | 0.3586 | +4.62%  |
|  8 | R2    | 1255.5  | 1266.1 | +0.85%  | 0.3438 | 0.3613 | +5.10%  |
|  8 | R3    | 1300.2  | 1200.6 | −7.66%  | 0.3466 | 0.3560 | +2.73%  |
|  4 | R1    |  709.1  |  708.0 | −0.15%  | 0.3475 | 0.3618 | +4.14%  |
|  4 | R2    |  694.5  |  701.5 | +1.01%  | 0.3422 | 0.3607 | +5.40%  |
|  4 | R3    |  708.6  |  675.1 | −4.72%  | 0.3457 | 0.3530 | +2.11%  |
|  2 | R1    |  390.1  |  380.3 | −2.51%  | 0.3420 | 0.3487 | +1.95%  |
|  2 | R2    |  390.0  |  379.8 | −2.61%  | 0.3489 | 0.3539 | +1.42%  |
|  2 | R3    |  390.4  |  392.7 | +0.59%  | 0.3434 | 0.3616 | +5.31%  |
|  1 | R1    |  205.5  |  205.7 | +0.07%  | 0.3429 | 0.3553 | +3.62%  |
|  1 | R2    |  202.5  |  203.4 | +0.48%  | 0.3530 | 0.3555 | +0.70%  |
|  1 | R3    |  204.7  |  201.0 | −1.84%  | 0.3511 | 0.3575 | +1.83%  |

</details>

### Qwen3-235B-A22B + EAGLE-3

| bs | TPS off | TPS on | ΔTPS    | AR off | AR on  | ΔAR      |
|---:|--------:|-------:|--------:|-------:|-------:|---------:|
| 16 | 1280.8  | 1337.3 | **+4.42%** | 0.3089 | 0.3510 | +13.62% |
|  8 |  777.3  |  805.9 | **+3.68%** | 0.3098 | 0.3502 | +13.03% |
|  4 |  469.0  |  492.4 | **+5.00%** | 0.3068 | 0.3539 | +15.36% |
|  2 |  278.0  |  288.8 | **+3.89%** | 0.3078 | 0.3519 | +14.32% |
|  1 |  154.1  |  159.7 | **+3.64%** | 0.3097 | 0.3488 | +12.62% |

### Observation

Acceptance rate improves across **all tested configurations** under non-greedy sampling (+2–4% on llama-70b, +12–15% on qwen-235b). The AR uplift translates to a TPS win on qwen-235b at every batch size, but not on llama-70b, where rejection sampling currently costs 0–9% TPS. Lower batch sizes also show enough run-to-run noise that some signs flip across rounds.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
Hi @mikeiovine please help to review this PR, thanks~

**Review by ruodil — APPROVED** [review]

**Comment by zhaoyangwang-nvidia** [comment]
Hi @NVIDIA/trt-llm-doc-owners @NVIDIA/trt-llm-llmapi-devs @NVIDIA/trt-llm-qa-function @NVIDIA/trt-llm-torch-models-devs @NVIDIA/trt-llm-torch-runtime-devs @NVIDIA/trt-llm-torch-spec-decoding please help to review this PR, thanks a lot.

**Review by jieli-matrix — APPROVED** [review]

File: tensorrt_llm/llmapi/llm_args.py [hunk 1]
```diff
@@ -896,20 +896,14 @@ class DecodingBaseConfig(StrictBaseModel):
...
         "PyTorch backend only.")
 
-    allow_advanced_sampling: bool = Field(
-        default=False,
-        status="prototype",
-        description=
-        "If true, allows non-greedy sampling when speculation is used. Only applicable "
-        "to 1-model code paths; non-greedy sampling is always enabled on 2-model paths."
-    )
-
```
Comment (mikeiovine) [inline tensorrt_llm/llmapi/llm_args.py:1130]: Since this an API breaking change, we need to leave the flag around for 3 months. I recommend keeping the flag and logging a warning that it will be removed soon if it is used. No further action is required as we are making sampling available by default.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 2]
```diff
@@ -201,19 +201,28 @@ def get_graph_key(
             self,
             batch: ScheduledRequests,
             new_tensors_device: Optional[SampleStateTensors] = None,
-            spec_resource_manager: Optional[BaseResourceManager] = None):
+            spec_resource_manager: Optional[BaseResourceManager] = None,
+            spec_metadata: Optional[Any] = None):
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: Let's try to avoid using `typing.Any`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -1104,31 +1071,61 @@ def _capture_generation_cuda_graphs(self,
...
-
-            for max_seq_len in max_seq_len_list:
-                warmup_request = self._create_cuda_graph_warmup_request(
-                    resource_manager, bs, draft_len, max_seq_len)
-                with self._release_batch_context(warmup_request,
-                                                 resource_manager) as batch:
-                    if batch is None:
-                        # No KV cache space, cannot continue capturing graphs
+        def _run_capture_pass(force_non_greedy: bool, label: str) -> None:
+            spec_metadata = getattr(self, 'spec_metadata', None)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: We can avoid `getattr` here. Just set `self.spec_metadata = None` when speculation is not in use

File: tensorrt_llm/_torch/speculative/eagle3.py [hunk 4]
```diff
@@ -811,27 +798,39 @@ def draft_decoder(
...
-
-        return draft_tokens
+        if spec_metadata is not None and batch_size is not None:
+            if (spec_metadata.use_rejection_sampling and draft_step is not None
+                    and not spec_metadata.is_all_greedy_sample):
+                return self._draft_sampler_advanced_for_rejection(
+                    logits, spec_metadata, batch_size, d2t, draft_step)
+            return self._draft_sampler_advanced(logits, spec_metadata,
+                                                batch_size, d2t)
+        return self._draft_sampler_greedy(logits, d2t)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/speculative/eagle3.py]: Does this branch ever get hit? Seems like spec_metadata and batch_size are always set to non-None

File: tensorrt_llm/llmapi/llm_args.py [hunk 5]
```diff
@@ -4148,12 +4143,12 @@ def validate_speculative_config(self):
...
-                                  Eagle3DecodingConfig):
-                    raise ValueError(
-                        "use_rejection_sampling is only supported for "
-                        "PyTorch Eagle3 one-model speculative decoding paths.")
+            if self.speculative_config.use_rejection_sampling and not isinstance(
+                    self.speculative_config, Eagle3DecodingConfig):
+                # Rejection sampling is only wired up for Eagle3 one-model paths.
+                # Silently fall back for other spec types so the new default
+                # (True) does not break them.
+                self.speculative_config.use_rejection_sampling = False
```
Comment (mikeiovine) [inline tensorrt_llm/llmapi/llm_args.py:4568]: Can we make this work with the other speculation types? It can be done in a followup.

Ideally the code should be unified so we don't have to make invasive changes for every single spec algorithm

**Review by mikeiovine — COMMENTED** [review]
Are any other spec algos always using greedy for draft sampling? That will need to be fixed too in a follow up

File: tensorrt_llm/llmapi/llm_args.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/llmapi/llm_args.py:1130]: Added

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py (hunk 2 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: Fixed

File: tensorrt_llm/_torch/speculative/eagle3.py (hunk 4 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/eagle3.py]: Good catch, the dead branch has been cleaned up

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Fixed

File: tensorrt_llm/llmapi/llm_args.py (hunk 5 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/llmapi/llm_args.py:4568]: Agreed. Any speculation type whose draft model emits per-token logits can use the same rejection path — the only requirement is p_draft(x) alongside the draft tokens. I will lift the dispatch into SpecMetadata and drop the whitelist guard in a followup PR.

**Review by mikeiovine — APPROVED** [review]

**Comment by zhaoyangwang-nvidia** [comment]
Hi @NVIDIA/trt-llm-doc-owners  please help to review this PR, thanks a lot~

**Review by QiJune — APPROVED** [review]
